ED/src/dynamics/farq_katul.f90

```diff
@@ -104,772 +111,676 @@ subroutine katul_lphys(can_prss,can_shv,can_co2,ipft,leaf_par,leaf_temp
+      !------------------------------------------------------------------------------------!
 
-            call fluxsolver(test_gsc, aero_resistance, can_co2, k1ci, k2ci, k3ci, k4ci, testci, testfc)
 
-            last_gV = test_gsc
-            
-            test_gV = test_gsc
-            test_fcV = testfc
-            test_ciV = testci
-          
-        else
-          ! canopy air is saturated...
-            test_ciV = can_co2
-            test_gV = 1000. / aero_resistance
-            call fluxsolver(test_gV, aero_resistance, can_co2, k1ci, k2ci, k3ci, k4ci, testci,test_fcV)
-        endif
 
 
-        ! 2. Repeat calculation for light-limited case
-        Jrate = Jrate * 0.25
-        a1gk = Jrate
-        a2gk = 2. * cp
-        k1ci = a1gk / can_co2 - Rdark / can_co2
-        k2ci = a1gk * aero_resistance / can_co2 - Rdark * aero_resistance / can_co2 - 1. + a2gk / can_co2
-        k3ci = -a1gk*cp/can_co2/can_co2-Rdark*a2gk/can_co2/can_co2
-        k4ci = -a1gk*cp/can_co2*aero_resistance/can_co2-Rdark*a2gk/can_co2*aero_resistance/can_co2-a2gk/can_co2
-    
-        if(leaf_vpr_prss > can_vpr_prss)then
-        ! Use newton's method to find the zero point of 
-        ! start with gsw from last time
-            if ((.not. isnan(last_gJ)) .and. last_gJ > 1e-10) then
-               test_gsc = last_gJ
-            else
-               test_gsc = cuticular_gsc 
-            endif
+      !------------------------------------------------------------------------------------!
+      !     Set Vm0 and find terms that typically depend upon Vm0 (Rd0, Jm0, TPm0).        !
+      ! This may be the default parameters, but in case trait plasticity is enabled, they  !
+      ! must be down-regulated.  Convert the resulting terms to mol/m2/s.                  !
+      !------------------------------------------------------------------------------------!
+      select case (trait_plasticity_scheme)
+      case (-2,-1,1,2,3)
+         !---------------------------------------------------------------------------------!
+         !      Within-canopy trait plasticity.  In this case, the input vm_bar is the     !
+         ! realized Vm0, and the input rd_bar is the realized Rd0, Jm0, and TPm0 are all   !
+         ! scaled with the vm_bar:Vm0 ratio.                                               !
+         !---------------------------------------------------------------------------------!
+         thispft(ib)%vm0  = dble(vm_bar) * umol_2_mol8
+         thispft(ib)%rd0  = dble(rd_bar) * umol_2_mol8
+         f_plastic8       = dble(vm_bar) / dble(vm0(ipft))
+         thispft(ib)%jm0  = f_plastic8 * dble(jm0 (ipft)) * umol_2_mol8
+         thispft(ib)%TPm0 = f_plastic8 * dble(TPm0(ipft)) * umol_2_mol8
+         !---------------------------------------------------------------------------------!
+      case default
+         !------ No plasticity: use default parameters. -----------------------------------!
```

Comment:
@xiangtaoxu I was checking this part of the code (and the equivalent one in farq_leuning.f90), and I think we should eliminate this select case block here and in farq_leuning.f90, and use only the "within-canopy" trait plasticity case.  In the case of light-controlled leaf phenology, vm_bar is not vm0, it is set by the site-level variable cpoly%vm_bar_toc(ico).  I thought I had changed this, but apparently I forgot to do it.

If both trait plasticity and light-controlled phenology are disabled, then vm_bar is equal to vm0, so it would be equivalent to the "case default".  If you and others are fine with this, I can submit a PR right after merging this.

Reply:
I think you are right. I have updated this pr accordingly for both farq_katul.f90 and farq_leuning.f90